File: src/go/pkg/netipc/service/raw/server_windows.go
//go:build windows

package raw

import (
	"sync"
	"sync/atomic"

	"github.com/netdata/netdata/go/plugins/pkg/netipc/protocol"
	windows "github.com/netdata/netdata/go/plugins/pkg/netipc/transport/windows"
)

// Server is an internal managed server bound to one expected request kind.
type Server struct {
	runDir                       string
	serviceName                  string
	config                       windows.ServerConfig
	expectedMethodCode           uint16
	handler                      DispatchHandler
	running                      atomic.Bool
	learnedRequestPayloadBytes   atomic.Uint32
	learnedResponsePayloadBytes  atomic.Uint32
	requestPayloadGrowthCeiling  uint32
	responsePayloadGrowthCeiling uint32
	nextSessionID                atomic.Uint64
	workerCount                  int
	wg                           sync.WaitGroup
	listenerMu                   sync.Mutex
	listener                     *windows.Listener // stored so Stop() can close it
}

// NewServer creates a new managed server.
func NewServer(
	runDir, serviceName string,
	config windows.ServerConfig,
	expectedMethodCode uint16,
	handler DispatchHandler,
) *Server {
	return NewServerWithWorkers(runDir, serviceName, config, expectedMethodCode, handler, defaultServerWorkerCount)
}

// NewServerWithWorkers creates a server with an explicit worker count limit.
func NewServerWithWorkers(
	runDir, serviceName string,
	config windows.ServerConfig,
	expectedMethodCode uint16,
	handler DispatchHandler,
	workerCount int,
) *Server {
	s := &Server{config: config}
	s.initCommon(
		runDir,
		serviceName,
		expectedMethodCode,
		handler,
		workerCount,
		config.MaxRequestPayloadBytes,
		config.MaxResponsePayloadBytes,
	)
	return s
}

type preparedWinShm struct {
	hybrid   *windows.WinShmContext
	busywait *windows.WinShmContext
}

func (p *preparedWinShm) take(profile uint32) *windows.WinShmContext {
	if p == nil {
		return nil
	}
	switch profile {
	case windows.WinShmProfileHybrid:
		ctx := p.hybrid
		p.hybrid = nil
		return ctx
	case windows.WinShmProfileBusywait:
		ctx := p.busywait
		p.busywait = nil
		return ctx
	default:
		return nil
	}
}

func (p *preparedWinShm) destroyAll() {
	if p == nil {
		return
	}
	if p.hybrid != nil {
		p.hybrid.WinShmDestroy()
		p.hybrid = nil
	}
	if p.busywait != nil {
		p.busywait.WinShmDestroy()
		p.busywait = nil
	}
}

const winShmProfiles = windows.WinShmProfileHybrid | windows.WinShmProfileBusywait

func (s *Server) prepareAcceptConfig() (uint64, windows.ServerConfig, *preparedWinShm, bool) {
	sessionID := s.nextSessionID.Add(1)
	cfg := s.config
	cfg.MaxRequestPayloadBytes = s.learnedRequestPayloadBytes.Load()
	cfg.MaxResponsePayloadBytes = s.learnedResponsePayloadBytes.Load()

	if cfg.SupportedProfiles&winShmProfiles == 0 {
		return sessionID, cfg, nil, true
	}

	prepared := &preparedWinShm{}
	for _, profile := range []uint32{windows.WinShmProfileHybrid, windows.WinShmProfileBusywait} {
		if cfg.SupportedProfiles&profile == 0 {
			continue
		}
		shm, err := windows.WinShmServerCreate(
			s.runDir, s.serviceName,
			cfg.AuthToken,
			sessionID,
			profile,
			cfg.MaxRequestPayloadBytes+uint32(protocol.HeaderSize),
			cfg.MaxResponsePayloadBytes+uint32(protocol.HeaderSize),
		)
		if err != nil {
			cfg.SupportedProfiles &^= profile
			cfg.PreferredProfiles &^= profile
			continue
		}
		if profile == windows.WinShmProfileHybrid {
			prepared.hybrid = shm
		} else {
			prepared.busywait = shm
		}
	}

	if cfg.SupportedProfiles == 0 {
		prepared.destroyAll()
		return sessionID, cfg, nil, false
	}

	if prepared.hybrid == nil && prepared.busywait == nil {
		return sessionID, cfg, nil, true
	}

	return sessionID, cfg, prepared, true
}

// Run starts the acceptor loop. Blocking.
func (s *Server) Run() error {
	listener, err := windows.Listen(s.runDir, s.serviceName, s.config)
	if err != nil {
		return err
	}
	s.setListener(listener)
	defer s.closeListener(listener)

	s.running.Store(true)
	sem := make(chan struct{}, s.workerCount)

	for s.running.Load() {
		sessionID, acceptCfg, preparedShm, ok := s.prepareAcceptConfig()
		if !ok {
			s.retryAcceptAfter(nil)
			continue
		}

		session, err := listener.AcceptWithConfig(sessionID, acceptCfg)
		if err != nil {
			if !s.retryAcceptAfter(func() {
				if preparedShm != nil {
					preparedShm.destroyAll()
				}
			}) {
				break
			}
			continue
		}

		if !s.acquireWorkerSlot(sem, func() {
			if preparedShm != nil {
				preparedShm.destroyAll()
			}
		}, session.Close) {
			continue
		}

		var shm *windows.WinShmContext
		if session.SelectedProfile == windows.WinShmProfileHybrid ||
			session.SelectedProfile == windows.WinShmProfileBusywait {
			shm = preparedShm.take(session.SelectedProfile)
			if shm == nil {
				if preparedShm != nil {
					preparedShm.destroyAll()
				}
				session.Close()
				<-sem
				continue
			}
		}
		if preparedShm != nil {
			preparedShm.destroyAll()
		}

		s.startSessionWorker(sem, func() {
			s.handleSession(session, shm)
		})
	}

	s.wg.Wait()
	return nil
}

// Stop signals the server to stop and unblocks Accept by closing the listener.
func (s *Server) Stop() {
	s.running.Store(false)
	s.closeListener(nil)
}

func (s *Server) setListener(listener *windows.Listener) {
	s.listenerMu.Lock()
	s.listener = listener
	s.listenerMu.Unlock()
}

func (s *Server) closeListener(listener *windows.Listener) {
	s.listenerMu.Lock()
	target := listener
	if target == nil {
		target = s.listener
	}
	if target != nil && s.listener == target {
		s.listener = nil
	}
	s.listenerMu.Unlock()

	if target != nil {
		target.Close()
	}
}

func (s *Server) handleSession(session *windows.Session, shm *windows.WinShmContext) {
	s.handleServerSession(serverSessionOps{
		maxRequestPayloadBytes:  session.MaxRequestPayloadBytes,
		maxResponsePayloadBytes: session.MaxResponsePayloadBytes,
		receive: func(recvBuf []byte) (protocol.Header, []byte, serverReceiveAction) {
			if shm != nil {
				mlen, err := shm.WinShmReceive(recvBuf, serverPollTimeoutMs)
				if err != nil {
					if err == windows.ErrWinShmTimeout {
						return protocol.Header{}, nil, serverReceiveContinue
					}
					return protocol.Header{}, nil, serverReceiveStop
				}
				if mlen < protocol.HeaderSize {
					return protocol.Header{}, nil, serverReceiveStop
				}
				hdr, err := protocol.DecodeHeader(recvBuf[:mlen])
				if err != nil {
					return protocol.Header{}, nil, serverReceiveStop
				}
				return hdr, recvBuf[protocol.HeaderSize:mlen], serverReceiveOK
			}

			ready, waitErr := session.WaitReadable(serverPollTimeoutMs)
			if waitErr != nil {
				return protocol.Header{}, nil, serverReceiveStop
			}
			if !ready {
				return protocol.Header{}, nil, serverReceiveContinue
			}

			hdr, payload, err := session.Receive(recvBuf)
			if err != nil {
				return protocol.Header{}, nil, serverReceiveStop
			}
			return hdr, payload, serverReceiveOK
		},
		send: func(respHdr *protocol.Header, payload []byte, msgBuf *[]byte) error {
			if shm == nil {
				return session.Send(respHdr, payload)
			}
			msg, err := serverEncodeSharedResponse(respHdr, payload, msgBuf)
			if err != nil {
				return err
			}
			return shm.WinShmSend(msg)
		},
		close: func() {
			if shm != nil {
				shm.WinShmDestroy()
			}
			session.Close()
		},
	})
}

File: src/crates/netipc/src/service/raw/server_windows.rs
use super::server::{ManagedServer, ServerConfig};
use super::server_session_windows::handle_session_win_threaded;
use crate::protocol::{NipcError, HEADER_SIZE};
use crate::transport::win_shm::{
    WinShmContext, PROFILE_BUSYWAIT as WIN_SHM_PROFILE_BUSYWAIT,
    PROFILE_HYBRID as WIN_SHM_PROFILE_HYBRID,
};
use crate::transport::windows::{NpListener, NpSession};
use std::sync::atomic::Ordering;

impl ManagedServer {
    /// Windows: run the acceptor loop over Named Pipes.
    pub fn run(&mut self) -> Result<(), NipcError> {
        let mut listener = NpListener::bind(
            &self.run_dir,
            &self.service_name,
            self.server_config.clone(),
        )
        .map_err(|_| NipcError::BadLayout)?;

        *self.listener_handle.lock().unwrap() = Some(listener.handle() as usize);

        self.running.store(true, Ordering::Release);

        let mut session_threads: Vec<std::thread::JoinHandle<()>> = Vec::new();

        while self.running.load(Ordering::Acquire) {
            let (session_id, accept_cfg, prepared_shm, ready) = self.prepare_windows_accept();
            if !ready {
                std::thread::sleep(std::time::Duration::from_millis(10));
                continue;
            }

            let session = match listener.accept_with_config(session_id, accept_cfg) {
                Ok(s) => s,
                Err(_) => {
                    if let Some(mut prepared) = prepared_shm {
                        prepared.destroy_all();
                    }
                    if !self.running.load(Ordering::Acquire) {
                        break;
                    }
                    std::thread::sleep(std::time::Duration::from_millis(10));
                    continue;
                }
            };

            session_threads.retain(|t| !t.is_finished());
            if session_threads.len() >= self.worker_count {
                if let Some(mut prepared) = prepared_shm {
                    prepared.destroy_all();
                }
                drop(session);
                continue;
            }

            let shm = self.finalize_windows_shm(&session, prepared_shm);
            if shm.is_none()
                && (session.selected_profile == WIN_SHM_PROFILE_HYBRID
                    || session.selected_profile == WIN_SHM_PROFILE_BUSYWAIT)
            {
                drop(session);
                continue;
            }

            let expected_method_code = self.expected_method_code;
            let handler = self.handler.clone();
            let running = self.running.clone();
            let learned_request_payload_bytes = self.learned_request_payload_bytes.clone();
            let learned_response_payload_bytes = self.learned_response_payload_bytes.clone();
            let t = std::thread::spawn(move || {
                handle_session_win_threaded(
                    session,
                    shm,
                    expected_method_code,
                    handler,
                    running,
                    learned_request_payload_bytes,
                    learned_response_payload_bytes,
                );
            });
            session_threads.push(t);
        }

        for t in session_threads {
            let _ = t.join();
        }

        Ok(())
    }

    fn prepare_windows_accept(&mut self) -> (u64, ServerConfig, Option<PreparedWinShm>, bool) {
        let session_id = self.next_session_id;
        self.next_session_id += 1;

        let mut cfg = self.server_config.clone();
        cfg.max_request_payload_bytes = self.learned_request_payload_bytes.load(Ordering::Acquire);
        cfg.max_response_payload_bytes =
            self.learned_response_payload_bytes.load(Ordering::Acquire);

        let shm_profiles =
            cfg.supported_profiles & (WIN_SHM_PROFILE_HYBRID | WIN_SHM_PROFILE_BUSYWAIT);
        if shm_profiles == 0 {
            return (session_id, cfg, None, true);
        }

        let mut prepared = PreparedWinShm::default();
        for profile in [WIN_SHM_PROFILE_HYBRID, WIN_SHM_PROFILE_BUSYWAIT] {
            if cfg.supported_profiles & profile == 0 {
                continue;
            }

            match WinShmContext::server_create(
                &self.run_dir,
                &self.service_name,
                self.server_config.auth_token,
                session_id,
                profile,
                cfg.max_request_payload_bytes + HEADER_SIZE as u32,
                cfg.max_response_payload_bytes + HEADER_SIZE as u32,
            ) {
                Ok(ctx) => prepared.insert(profile, ctx),
                Err(_) => {
                    cfg.supported_profiles &= !profile;
                    cfg.preferred_profiles &= !profile;
                }
            }
        }

        if cfg.supported_profiles == 0 {
            prepared.destroy_all();
            return (session_id, cfg, None, false);
        }

        if prepared.is_empty() {
            return (session_id, cfg, None, true);
        }

        (session_id, cfg, Some(prepared), true)
    }

    fn finalize_windows_shm(
        &self,
        session: &NpSession,
        mut prepared: Option<PreparedWinShm>,
    ) -> Option<WinShmContext> {
        let profile = session.selected_profile;
        if profile != WIN_SHM_PROFILE_HYBRID && profile != WIN_SHM_PROFILE_BUSYWAIT {
            if let Some(ref mut prepared) = prepared {
                prepared.destroy_all();
            }
            return None;
        }
        let mut prepared = prepared?;
        // Keep the negotiated context and destroy every unused prepared context.
        let selected = prepared.take(profile);
        prepared.destroy_all();
        selected
    }
}

#[derive(Default)]
struct PreparedWinShm {
    hybrid: Option<WinShmContext>,
    busywait: Option<WinShmContext>,
}

impl PreparedWinShm {
    fn insert(&mut self, profile: u32, ctx: WinShmContext) {
        if profile == WIN_SHM_PROFILE_HYBRID {
            self.hybrid = Some(ctx);
        } else if profile == WIN_SHM_PROFILE_BUSYWAIT {
            self.busywait = Some(ctx);
        }
    }

    fn take(&mut self, profile: u32) -> Option<WinShmContext> {
        if profile == WIN_SHM_PROFILE_HYBRID {
            self.hybrid.take()
        } else if profile == WIN_SHM_PROFILE_BUSYWAIT {
            self.busywait.take()
        } else {
            None
        }
    }

    fn destroy_all(&mut self) {
        if let Some(mut ctx) = self.hybrid.take() {
            ctx.destroy();
        }
        if let Some(mut ctx) = self.busywait.take() {
            ctx.destroy();
        }
    }

    fn is_empty(&self) -> bool {
        self.hybrid.is_none() && self.busywait.is_none()
    }
}
